ibm_i/datadog_checks/ibm_i/check.py

```diff
@@ -78,17 +70,134 @@ def check(self, _):
                 hostname=self._query_manager.hostname,
             )
 
-    def execute_query(self, query):
-        # https://github.com/mkleehammer/pyodbc/wiki/Connection#execute
-        with closing(self.connection.execute(query)) as cursor:
+    def cancel(self):
+        # When the check gets cancelled, clean up the connection subprocess.
+        self._delete_connection_subprocess()
+
+    def handle_query_error(self, error):
+        self._current_errors += 1
+        return error
+
+    @property
+    def connection_subprocess(self):
+        if self._subprocess is None:
+            self._create_connection_subprocess()
+        return self._subprocess
+
+    def _create_connection_subprocess(self):
+        self._subprocess = subprocess.Popen(
+            # TODO: Change to sys.executable once we're based on Agent 7.30.0
+            [
+                "/opt/datadog-agent/embedded/bin/python3",
+                "-c",
+                "from datadog_checks.ibm_i.query_script import query; query()",
+            ],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            text=True,
+        )
+
+        # Set stdout reader as non-blocking, we don't want to
+        # block .read() calls to be able to time out.
+        fl = fcntl.fcntl(self._subprocess.stdout.fileno(), fcntl.F_GETFL)
+        fcntl.fcntl(self._subprocess.stdout, fcntl.F_SETFL, fl | os.O_NONBLOCK)
+
+        # Set stderr reader as non-blocking, we don't want to
+        # wait until EOF is sent, we only want to read whatever is there when
+        # we try to return errors.
+        fl = fcntl.fcntl(self._subprocess.stderr.fileno(), fcntl.F_GETFL)
+        fcntl.fcntl(self._subprocess.stderr, fcntl.F_SETFL, fl | os.O_NONBLOCK)
```

Comment:
I guess this is a workaround to the warning here https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stderr ? Is this better than using `Popen.communicate` in some way?

Reply:
[`Popen.communicate`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate) waits for the process to terminate:

> Interact with process: Send data to stdin. Read data from stdout and stderr, until end-of-file is reached. Wait for process to terminate and set the returncode attribute. 

so it's not really usable here (we don't want to kill the subprocess after communicating with it once, since we want it to keep the connection alive).

Reply:
I see, makes sense